docs/src/pages/ThemingSpecification.mdx

```diff
@@ -0,0 +1,185 @@
+})
+```
+
+#### CSSinJS usage
+
+If users decide to use CSSinJS we have to propose some structure of how they can define all things required for theming their components.
+
+I will describe one proposal of how the theme object can look like:
+
+##### Global tokens
+
+These are tokens that can be reused in many components, things like density, font sizes, font families, colors, shadows etc. Having these global tokens, can help designers to have more consistent look and feel on the components which are part of one application/theme.
+
+The global tokens can be defined as a plain javascript object. We can offer some utilities for generating color palettes if there is an agreement of how those should be generated, but anyway, the global tokens should be defined by the users in any format they would like to structure them, and our framework should not depend on something that is defined (or not defined) there.
+
+##### Component tokens
+
+These are tokens which are specific for a component. They can use some of the global tokens for their values.
+
+**Open for a debate here**
+
+We have the components tokens in Teams theme and those were shown as a good abstraction. On the other hand, people overused them for adding inline overrides, because these tokens can be overridden on any instance of a component. So, maybe the best option will be to restrict the ways of where these tokens can be defined (only in the theme and not on instance level), so we can avoid similar misusages in future.
+
+##### Component styles
+
+The component styles should be functions that return CSSinJS definition that is compatible with the CSSinJS engine used. The input parameters of these functions are:
+
+- **design terms** - in the react implementation, these are basically subset of the component's properties. It is important to note here that for better performance we need to restrict these to be primitive values like: booleans, strings, numbers, so we can easily hash based on them.
+- **design tokens** - these are basically the component's tokens. If we support inline overrides on these tokens, it will have some performance degradation
+
+As we mention before, parents component should be able to specify styles for their slots. We can have two ways of how this can be implemented
+
+- define slots' styles inside the styles of the parent component
+
+```
+const buttonStyles = {
+  root: ({ terms, tokens }) => ({
+    // some css in js definition
+  }),
+  icon: ({ terms, tokens }) => ({
+    // some css in js definition
+  }),
+}
+```
+
+- create composed components for the slots if they require some additional styles
+
+```
+const buttonStyles = {
+  root: ({ terms, tokens }) => ({
+    // some css in js definition
+  }),
+}
+
+const buttonIconStyles = {
+  root: ({ terms, tokens }) => ({
+    // some css in js definition
+  }),
+}
+```
+
+There are props and cons to both approaches, so I will just try to list them all and we can discuss them:
```

Comment:
This is a good list.

2 comments in favor of creating composed components:
- providing styling from a parent requires knowledge of the internals of another component, which creates coupling
- using a composed component makes it easy to swap in another component which might want all its own styles.